src/azul/service/manifest_service.py

```diff
@@ -815,15 +824,39 @@ def manifest_config(self) -> ManifestConfig:
                               combine_script='return new ArrayList(params._agg.fields)',
                               reduce_script=reduce_script)
         es_search = es_search.extra(size=0)
-        response = es_search.execute()
-        assert len(response.hits) == 0
+        try:
+            response = es_search.execute()
+        # If timeout is exceeded the request is partitioned to reduce load on ES
+        except ConnectionTimeout:
+            fields = self._partition_search(es_search)
+        else:
+            assert len(response.hits) == 0
+            fields = response.aggregations.fields.value
         return {
             'contents': {
                 value: value.split('.')[-1]
-                for value in sorted(response.aggregations.fields.value)
+                for value in sorted(fields)
             }
         }
 
+    def _partition_search(self, es_search: Search) -> Set[str]:
+        requests = [es_search.query(Q('bool',
+                                      must=Q('prefix',
+                                             **{'bundles.uuid.keyword': p})))
+                    for p in map(''.join, product('0123456789abcdef'))]
+        logger.info('Sending asynchronous requests to Elasticsearch for manifest config')
+        start = time.time()
+        with ThreadPoolExecutor(max_workers=(len(requests) // 2)) as tpe:
+            fields = list(tpe.map(self._get_partition_search_fields, requests))
+        logger.info('Manifest config requests completed after %.003fs',
+                    time.time() - start)
+        return set(chain(*fields))
+
+    def _get_partition_search_fields(self, es_search: Search) -> List[str]:
+        response = es_search.execute()
+        assert len(response.hits) == 0
```

Comment:
what happens if there is a single thread that fails this assertion? Can the assertion be done at a later stage? Should we log the AssertionFailure? What does the user see?

Reply:
Since the [request size gets set](https://github.com/DataBiosphere/azul/commit/fec0e3274136149c9978c681012750e36d7ea039#diff-0fbaa7c358bbfb13d3626f7bf3d5ad051f6cde5cf88c646080cb2b3ea99818e0R825) before the prefix query is prepared, it is unlikely that a single thread fails the assertion. In any case, if one fails they should all fail, otherwise they would likely generate an incomplete manifest header row. 
I don't think the assertion can happen later since only the aggregate result of the Elasticsearch query is returned, the rest of the response (including hits) is disregarded. 
Wouldn't the assertion failure come up in the logs if a job fails? What was your resolution from the time a failure was being triggered inside a TPE?
The user simply sees a failed attempt at generating a manifest. This would only be accesible via the logs of the step-function running the job, which I believe is ok.